### ISSUE TITLE
Iterator to ParallelIterator adapter

### ISSUE THREAD

**Issue by Michael-F-Bryan** [nikomatsakis/rayon#355]
**TL; DR:** Is there a way to turn any generic `std::iter::Iterator` into a `ParallelIterator`? What I'm thinking of is a sort of adapter which will take an iterator (or anything which knows how to generate the next element, really), and then allow you to do all future operations on the elements in parallel.

As a bit more context, I'm using the [`iproduct!()`][ip] macro to get an iterator over all the possible coordinates in an image given its width and height. Then for each pixel I'm doing a fairly expensive computation (generating fractals).

(rough pseudocode)
```rust
#[macro_use]
extern crate itertools;

let pixels = iproduct!(0..height, 0..width)
  .map(|coord| (coord, calculate_fractal(coord)));

generate_image_from_pixels(pixels);
```

I've got multiple cores on my machine, so the obvious thing to try is using rayon to get parallelism "for free". 

```rust
use rayon::prelude::*;

let pixels = iproduct!(0..height, 0..width)
  .into_par_iter()
  .map(|coord| (coord, calculate_fractal(coord)));
```

This obviously fails because [`itertools::structs::Product`][product] (created by the `iproduct!()` macro) doesn't implement `IntoParallelIterator`. Are there plans for some sort of adapter which could enable you to run jobs in parallel on the items generated from a normal `Iterator`? 

[ip]: https://docs.rs/itertools/0.6.0/itertools/macro.iproduct.html
[product]: https://docs.rs/itertools/0.6.0/itertools/structs/struct.Product.html

**Comment by cuviper** [comment]
In general, that's #46.  For this particular example, you could fairly easily open-code the parallel product with a `flat_map` or nested `for_each` on parallel ranges.

We've also been talking about implementing some itertools stuff in parallel, so perhaps we'll have a Rayon `Product`.

**Comment by Michael-F-Bryan** [comment]
Good point, I'll close this issue in favour of #46. I noticed that #312 also *seems* to be a solution, @cuviper is there anything I can do to help get it merged?

**Comment by cuviper** [comment]
Basically I think that PR is just forgotten, never taken past the "WIP" stage.  You can start testing to see how well it works in your situation, and reporting your results may rekindle interest. :)

**Comment by saolof** [comment]
Is there any conversion to a normal iterator that feeds into a buffered iterator, which is not necessarily thread safe, but lets you handle those pesky FFI calls which are not thread safe?

**Comment by oovm** [comment]
Is there any solution for `iproduct!`?
